Petsc bddc preconditioner

nabw opened the pull request:
This fork adds support for substructuring domain decomposition preconditioners in PETSc through the allocation of an IS matrix. This has been further expanded with the BDDC preconditioner. More specifically, I added the function

`reinit_IS`

which allocates the local matrix, together with the internal `do_reinit_IS` functions. After this, an interface to the BDDC preconditioner was added in the standard way, with the most relevant parameters exposed to the user. As usual, further options can be set, and I have indeed verified it does so, through the -options_file FILE flag from the executable. 

Please let me know if you would like for this branch to include tests for validation.

drwells commented:
This is a nice feature and its the kind of thing we should include in the library. I have a few high-level comments:
1. Despite the extra copy, since we are passing in node locations we should be providing a `std::vector<Point<spacedim>>` here to match the APIs used elsewhere in the library. This class can then convert it into `std::vector<PetscReal>` internally.
2. We need some extra documentation on edge and face variables (and what is meant by primal or dual here). I think this terminology comes from staggered FDM grids and it is not a term we use in deal.II.
3. We do need to incorporate some basic unit tests into the library, like what we have for the other PETSc preconditioners. If you already have some test code that runs at low processor counts this should be easy to do.
4. All PETSc calls should have return codes checked with ` AssertThrow(ierr == 0, ExcPETScError(ierr));`

nabw commented:
First of all, thank you for reviewing this. I will include your comments as soon as possible. Let me further comment only on the second one: 

> 2. We need some extra documentation on edge and face variables (and what is meant by primal or dual here). I think this terminology comes from staggered FDM grids and it is not a term we use in deal.II.

The vertices, edges, faces come from some algebraic rules pre-defined internally, meant to follow the intuition one has on a cube. For example, a face dof would be one that is shared by exactly two processors. The primal and dual part is strictly related to domain decomposition, so that primal dofs are the ones where there is continuity in the different domains, dual ones instead are "repeated" as they do not take the same value in each sharing process. I will add some references and explain it better for clarification (starting from this paragraph!)

Keep you updated.

peterrum commented on source/lac/petsc_precondition.cc:
```diff
@@ -850,6 +850,105 @@ namespace PETScWrappers
+    const unsigned int            coords_cdim_,
+    const types::global_dof_index coords_n_,
+    const PetscReal *             coords_data_)
+    : use_vertices(use_vertices_)
```
```suggestion
    : use_vertices(use_vertices)
```

peterrum reviewed (commented):
@nabw Could you add a test so that it is clear how to use the new functions/classes!

nabw commented on include/deal.II/lac/petsc_precondition.h:
```diff
@@ -957,6 +957,111 @@ namespace PETScWrappers
+   * PETSc.
+   *
+   * @ingroup PETScWrappers
+   * @author Nicolas Barnafi, 2022
```
Thank you Peter. I added the corresponding minor change file.

nabw commented on include/deal.II/lac/petsc_precondition.h:
```diff
@@ -957,6 +957,111 @@ namespace PETScWrappers
 
+  /**
+   * A class that implements the interface to use the BDDC preconditioner from
+   * PETSc.
```
Finally starting this again. Added comments on the way the preconditioner works, plus the PETSc class and the main reference.

nabw commented on include/deal.II/lac/petsc_precondition.h:
```diff
@@ -957,6 +957,111 @@ namespace PETScWrappers
+      /**
+       * Support for H1 problems
+       */
+      unsigned int            coords_cdim;
+      types::global_dof_index coords_n;
+      const PetscReal
+        *coords_data; /* not referenced, consumed at initialize time */
```
Missing comments, added one for each field.

nabw commented on include/deal.II/lac/petsc_sparse_matrix.h:
```diff
@@ -498,6 +501,24 @@ namespace PETScWrappers
+       */
+      template <typename SparsityPatternType>
+      void
+      reinit_IS(const IndexSet &           local_rows,
+                const IndexSet &           local_active_rows,
```
Yes, exactly that. Active is a bit smaller than the 'relevant' ones, and suffices for this type of preconditioner. They are obtained from DoFTools, and I added a small comment on the comments above the definition.

nabw commented on include/deal.II/lac/petsc_sparse_matrix.h:
```diff
@@ -498,6 +501,24 @@ namespace PETScWrappers
+                const IndexSet &           local_active_rows,
+                const IndexSet &           local_columns,
+                const IndexSet &           local_active_columns,
+                const SparsityPatternType &sparsity_pattern,
```
It provides the row length and some useful iterators for the do_reinit_IS function.

nabw commented on include/deal.II/lac/petsc_sparse_matrix.h:
```diff
@@ -613,6 +634,17 @@ namespace PETScWrappers
                 const SparsityPatternType &sparsity_pattern);
 
+      /**
+       * Same as previous functions, but here we consider active dofs for MATIS.
```
it is a PETSc matrix type for substructuring preconditioenrs. I rephrased it for clarity.

nabw commented on source/lac/petsc_parallel_sparse_matrix.cc:
```diff
@@ -406,6 +428,203 @@ namespace PETScWrappers
+                      &is_glob_row);
+      ISLocalToGlobalMappingCreateIS(is_glob_row, &l2gmap_row);
+      ISDestroy(&is_glob_row);
+      ISLocalToGlobalMappingViewFromOptions(l2gmap_row, NULL, "-view_map");
```
Is it ok if I just add a comment for the function to state they do the same thing? In the end the important object is the mapping and not the indices.

nabw commented on source/lac/petsc_parallel_sparse_matrix.cc:
```diff
@@ -406,6 +428,203 @@ namespace PETScWrappers
+      // class.
+      // if (preset_nonzero_locations == true)
+
+      // In the MATIS case, we use the local matrix instead
```
This is strictly related to the matrix type, which is stated after creating the matrix with

ierr = MatCreateIS(...);

nabw commented on source/lac/petsc_parallel_sparse_matrix.cc:
```diff
@@ -406,6 +428,203 @@ namespace PETScWrappers
+
+      IS is_glob_row, is_glob_col;
+      // Create row index set
+      ISLocalToGlobalMapping l2gmap_row;
```
Sorry, I don't know what you mean by this. the l2gmaps are usally objects looking like

[0, ... , n_local_dofs)

which are used for extending local manipulations to the global matrix.

nabw commented on include/deal.II/lac/petsc_precondition.h:
```diff
@@ -957,6 +957,111 @@ namespace PETScWrappers
   };
 
+  /**
+   * A class that implements the interface to use the BDDC preconditioner from
```
I have added a minimal helmholtz (good sign) test that shows this, perhaps it could be referenced from the code?

peterrum commented:
@nabw I'll give the PR another round in the next days.

drwells commented on include/deal.II/lac/petsc_precondition.h:
```diff
@@ -957,6 +957,131 @@ namespace PETScWrappers
+                     const bool                    symmetric    = false,
+                     const unsigned int            coords_cdim  = 0,
+                     const types::global_dof_index coords_n     = 0,
+                     const PetscReal *             coords_data  = nullptr);
```
This was mentioned in at least one other place, but to match the convention used elsewhere in the library this really should be a `std::vector<Point<spacedim>>` vector. Avoiding templates is good in general though this (like the cgal additional data struct) is a place where we should use them.

nabw replied:
I understand, I am working on this now. 

An extreme side note, but I have taken long to start because my reinstallation has been a bit difficult due to the migration of METIS to github only. I have some workarounds for candi, if you every face the problem.

drwells replied:
I usually just let PETSc download and install METIS for me to avoid these problems.

drwells commented on include/deal.II/lac/petsc_precondition.h:
```diff
@@ -957,6 +957,131 @@ namespace PETScWrappers
+       * Set the number of coordinates that each DoF has, i.e. 2 in 2D and 3 in
+       * 3D.
+       */
+      unsigned int coords_cdim;
```
Same thing here - this should just be a compile-time constant (spacedim).

nabw replied:
Included in previous mod

drwells commented on include/deal.II/lac/petsc_precondition.h:
```diff
@@ -957,6 +957,131 @@ namespace PETScWrappers
+      /**
+       * Set the number of degrees of freedom that the problem has.
+       */
+      types::global_dof_index coords_n;
```
Is this the same as the number of coordinates?

nabw replied:
Included in previous mod (yes)

drwells commented on include/deal.II/lac/petsc_sparse_matrix.h:
```diff
@@ -27,6 +27,9 @@
 
 #    include <vector>
 
+#    include "petscis.h"
+#    include "petscistypes.h"
```
I'm surprised clang-format accepted this - its probably because of the quotes. Could you switch to using brackets with these inclusions? That should fix it.

nabw replied:
done!

drwells commented on include/deal.II/lac/petsc_sparse_matrix.h:
```diff
@@ -498,6 +501,24 @@ namespace PETScWrappers
+       */
+      template <typename SparsityPatternType>
+      void
+      reinit_IS(const IndexSet &           local_rows,
```
```suggestion
      reinit(const IndexSet &           local_rows,
```
The other functions also take index sets so lets just reuse the same name - our convention is that we always call this `reinit()`.

nabw replied:
Ok. Is there a good place to put some kind of warning due to the different underlying matrix type? 

If not, it is always possible to just assume that the user knows what they are doing. Surely not always the case for me 😅, that's why I kept it separate.

drwells replied:
The right answer here is to document the requirement and use some combination of `MatGetType()` and our own `AssertThrow()` macro to guarantee that the correct `Mat` type is present.

drwells commented on include/deal.II/lac/petsc_sparse_matrix.h:
```diff
@@ -613,6 +634,18 @@ namespace PETScWrappers
+       */
+      template <typename SparsityPatternType>
+      void
+      do_reinit_IS(const IndexSet &           local_rows,
```
```suggestion
      do_reinit(const IndexSet &           local_rows,
```
same thing.

drwells commented on source/lac/petsc_parallel_sparse_matrix.cc:
```diff
@@ -85,6 +85,28 @@ namespace PETScWrappers
+
+      // get rid of old matrix and generate a new one
+      const PetscErrorCode ierr = destroy_matrix(matrix);
+      AssertThrow(ierr == 0, ExcPETScError(ierr));
```
We haven't done it yet but `destroy_matrix()` should be deprecated at some point before the next release since it isn't necessary anymore (we don't support versions of PETSc which had a different destruction API), so it is better to use `MatDestroy` directly.

nabw replied:
Changed it as suggested

drwells commented on source/lac/petsc_parallel_sparse_matrix.cc:
```diff
@@ -406,6 +428,218 @@ namespace PETScWrappers
+      IndexSet::size_type n_local_active_rows = local_active_rows.n_elements();
+      IndexSet::size_type n_local_active_cols =
+        local_active_columns.n_elements();
+      std::vector<int> idx_glob_row(n_local_active_rows);
+      std::vector<int> idx_glob_col(n_local_active_cols);
```
```suggestion
      std::vector<PetscInt> idx_glob_row(n_local_active_rows);
      std::vector<PetscInt> idx_glob_col(n_local_active_cols);
```
(since these are passed to a PETSc function)

nabw replied:
Changed it as suggested

drwells commented on source/lac/petsc_parallel_sparse_matrix.cc:
```diff
@@ -406,6 +428,218 @@ namespace PETScWrappers
+        local_active_columns.n_elements();
+      std::vector<int> idx_glob_row(n_local_active_rows);
+      std::vector<int> idx_glob_col(n_local_active_cols);
+      unsigned int     k;
```
We can just define this in the loop itself - it should also be `IndexSet::size_type`.

nabw replied:
Changed it as suggested

drwells commented on tests/petsc/bddc.cc:
```diff
@@ -0,0 +1,171 @@
+//
+// ---------------------------------------------------------------------
+
+// Test that it is possible to instantiate a LinearOperator object for all
+// different kinds of PETSc matrices and vectors
+// TODO: A bit more tests...
```
We need a new test summary :)

nabw replied:
Hahah sorry, dumb mistake. It is correct now.

drwells reviewed (changes requested):
We need to change a few minor things.

Another thing that needs to be added is a test which uses the DoF coordinates. It would also help if we added some additional documentation explaining how to use something like `FE_Q<dim>(2)` in parallel with this preconditioner.

drwells commented:
Sorry its taken us so long to get back to this! This is a cool preconditioner and a nice feature for the library.

nabw commented:
Ok, I just finished adding everything. @drwells regarding the points, I added them directly to the existing test, as the required modification mas minimal. For FEQ(2) there is no modification required, the preconditioner will run perfectly well in that scenario :) I have actually tested this, and added some details on the usage of this interface on a preprint (https://arxiv.org/pdf/2208.06191.pdf), but I wasn't sure about how to reference it. Is it possible?

drwells commented:
Yes - we can add a new reference to `doc/doxygen/references.bib` and cite it in the normal way in the documentation.

drwells commented:
Yes - we can add a new reference to `doc/doxygen/references.bib` and cite it in the normal way in the documentation.

drwells commented on doc/news/changes/minor/20221005NicolasBarnafi:
```diff
@@ -0,0 +1,3 @@
+New: The PetscWrappers::PreconditionBDDC preconditioner has been implemented. It requires a special matrix type (IS), which is initialized with the new function PetscWrappers::reinit_IS.
```
we need to update this too since we renamed `reinit_IS` to `reinit`.

drwells commented on include/deal.II/lac/petsc_precondition.h:
```diff
@@ -957,14 +958,129 @@ namespace PETScWrappers
 
-
+template class PETScWrappers::PreconditionBDDC<2>;
+template class PETScWrappers::PreconditionBDDC<3>;
```
These explicit instantiations should be in a source file, not the header.

nabw commented:
Ok, finished adding last comments :)

drwells commented:
/rebuild

nabw commented:
Thanks @drwells  @peterrum  for all the help!

drwells commented:
It looks like the CI found a small problem - this isn't compatible with PETSc 3.7 (the oldest version of PETSc we support). Can you add some version checks?

nabw commented:
> It looks like the CI found a small problem - this isn't compatible with PETSc 3.7 (the oldest version of PETSc we support). Can you add some version checks?

Sure. Do you mean to update the oldest supported version or to add preproc directives to load this preconditioner only is a sufficiently recent version is used?

drwells commented:
We should only enable this preconditioner when the detected version of PETSc is new enough, e.g., put
```cpp
#if DEAL_II_PETSC_VERSION_GTE(a,b,c)
// code for the normal thing
#else
AssertThrow(false, ExcMessage("This preconditioner is only available with PETSc a.b.c or newer");
#endif
```
around all the class members.

drwells commented:
My mistake - I forgot to explain the test part.

To get the correct version check in the test you should not use the preprocessor version check but instead rename the output file to
    bddc.with_petsc.geq.3.9.mpirun=2.output

that way we don't compile (or run) the test with older versions of PETSc.

nabw commented:
> My mistake - I forgot to explain the test part.
> 
> To get the correct version check in the test you should not use the preprocessor version check but instead rename the output file to
> 
> ```
> bddc.with_petsc.geq.3.9.mpirun=2.output
> ```
> 
> that way we don't compile (or run) the test with older versions of PETSc.

hahah I opened the browser for this... Thanks, working on it right now

nabw commented:
> My mistake - I forgot to explain the test part.
> 
> To get the correct version check in the test you should not use the preprocessor version check but instead rename the output file to
> 
> ```
> bddc.with_petsc.geq.3.9.mpirun=2.output
> ```
> 
> that way we don't compile (or run) the test with older versions of PETSc.

Ok, there are two small details on this: 
-) 3.10 is required for the Options commands used with the L2G mappings in reinit. Without reading from options, actually 3.9 should be enough
-) Even more precisely, 3.7 should suffice in most cases, unless the coordinates option is used, then at least 3.9 is required.

I thought that those were unnecessarily complicated if statements, so I just kept the biggest one. Seems reasonable to me, but I leave the choice to you.

drwells commented:
I think we should use 3.10 since it supports everything.

drwells commented:
There are still some unused variable warnings that need to be fixed. While this patch is going on we are cleaning up our headers so it looks like you need to rebase too - let me know if you need a hand with either part.

nabw commented:
> There are still some unused variable warnings that need to be fixed. While this patch is going on we are cleaning up our headers so it looks like you need to rebase too - let me know if you need a hand with either part.

Thanks for the help, I remerged with master but found out that the problem was ocurring because whenever an older PETSc version was being used, the preprocessor directive we used was leaving some unused variables only on that case. I added some casts to void for this, it should work now (hopefully!)

nabw commented:
@drwells everything fine now! Except for some errors on a windows file upload. Not sure what it is, nor if I can do something about it.

drwells commented:
Those failures are spurious - I think we just fixed that in our CI.
